libsolidity/analysis/TypeChecker.cpp

```diff
@@ -2078,6 +2086,100 @@ void TypeChecker::typeCheckABIEncodeFunctions(
 	}
 }
 
+void TypeChecker::typeCheckABIEncodeCallFunction(FunctionCall const& _functionCall)
+{
+	vector<ASTPointer<Expression const>> const& arguments = _functionCall.arguments();
+
+	// Expecting first argument to be the function pointer and second to be a tuple.
+	if (arguments.size() != 2)
+	{
+		m_errorReporter.typeError(
+			6219_error,
+			_functionCall.location(),
+			"Expected two arguments: a function pointer followed by a tuple."
+		);
+		return;
+	}
+
+	auto const functionPointerType = dynamic_cast<FunctionTypePointer>(type(*arguments.front()));
+
+	if (!functionPointerType)
+	{
+		m_errorReporter.typeError(
+			5511_error,
+			arguments.front()->location(),
+			"Expected first argument to be a function pointer, not \"" +
+			type(*arguments.front())->canonicalName() +
+			"\"."
+		);
+		return;
+	}
+
+	if (functionPointerType->kind() != FunctionType::Kind::External)
+	{
+		SecondarySourceLocation ssl{};
+
+		if (functionPointerType->hasDeclaration())
+			ssl.append("Function is declared here:", functionPointerType->declaration().location());
+		m_errorReporter.typeError(
+			3509_error,
+			arguments[0]->location(),
+			ssl,
+			"Function must be \"public\" or \"external\". Did you forget to prefix \"this.\"?"
+		);
+		return;
+	}
+
+	solAssert(!functionPointerType->takesArbitraryParameters(), "No checks possible for arbitrary parameters.");
+
+	// Tuples with only one component become that component
+	auto const argumentTuple = std::dynamic_pointer_cast<TupleExpression const>(arguments[1]);
```

Comment:
You could also use
```
	vector<ASTPointer<Expression>> arguments;
	if (auto const* argumentTuple = dynamic_cast<TupleExpression coonst*>(arguments[1].get()))
		arguments = argumentTuple->components();
	else
		arguments = {arguments[1]};
```
And then you can use `arguments.size()` instead of `numArguments` and `getArg` below is just `arguments.at(i)`.

Reply:
I think I tried something like that, but the problem ended up being that `components()` returns an array without any const, where as `_functionCall.arguments()` is an array of const expressions. 

I can't assign/add const `arguments[1]` to `arguments` and if I make `arguments` contain const expressions then I can't assign `components()` to `arguments`..

```
	// Tuples with only one component become that component
	vector<ASTPointer<Expression>> callArguments;

	if (auto const* argumentTuple = dynamic_cast<TupleExpression const*>(arguments[1].get()))
		callArguments = argumentTuple->components(); // failure here if you change `callArguments` to have Expression const as element type, but below works
	else
		callArguments.push_back(arguments[1]); // failure here, const to no-const conversion
```

Reply:
You can use const pointers and then convert the non-const to the const-version using `| ranges::to<decltype(callArguments)>;` or something.
another way is to use `{argumentTuple->components().begin(), argumentTuple->components().end()}`